### PR TITLE
feat(routes): display the appropriateness message

### PR DESCRIPTION
**Pull request by rBergCode**
# Summary
The goal was to have a message displaying and telling the user if the hike is suitable or not for him.

# Details
As I added this feature, multiple things had to be done:
- Needed for the profile informations in `MapScreen` and `HikeDetailScreen`
- Needed to fix tests using those screens (as they now need a valid profile
- Needed to fix end-to-end 2 test as it was not creating a profile and still going to map screen

### CODE REVIEW

File: app/src/androidTest/java/ch/hikemate/app/endtoend/EndToEndTest2.kt [hunk 1]
```diff
@@ -47,8 +48,6 @@ class EndToEndTest2 {
     val context = ApplicationProvider.getApplicationContext<Context>()
     FirebaseApp.initializeApp(context)
 
-    auth.createUserWithEmailAndPassword(email, password)
```
Comment (mchac15) [inline app/src/androidTest/java/ch/hikemate/app/endtoend/EndToEndTest2.kt:50]: Question: Why was this removed?

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt [hunk 2]
```diff
@@ -180,28 +192,58 @@ fun HikeDetailScreen(
...
-        onZoomOut = { mapView.controller.zoomOut() },
-        modifier =
-            Modifier.align(Alignment.BottomEnd)
-                .padding(bottom = MapScreen.BOTTOM_SHEET_SCAFFOLD_MID_HEIGHT + 8.dp))
-
-    // Hike Details bottom sheet
-    HikeDetails(detailedRoute, savedHikesViewModel, elevationData)
+  LaunchedEffect(Unit) {
+    if (authViewModel.currentUser.value == null) {
+      Log.e("MapScreen", "User is not signed in")
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:197]: Nitpick: Shouldn't the log message say `HikeDetailScreen` here?

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt [hunk 3]
```diff
@@ -363,64 +393,90 @@ fun MapScreen(
...
-                onClick = {
-                  MapScreen.launchSearch(isSearching, hikingRoutesViewModel, mapView, context)
-                },
-                modifier =
-                    Modifier.align(Alignment.BottomCenter)
-                        .padding(bottom = MapScreen.BOTTOM_SHEET_SCAFFOLD_MID_HEIGHT + 8.dp))
+  val errorMessageIdState = profileViewModel.errorMessageId.collectAsState()
+  val profileState = profileViewModel.profile.collectAsState()
+
+  when {
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:399]: Question: Since you are reusing this pattern twice on the MapScreen aswell as in the HikeDetailsScreen shouldn't this be made into a function? Or is that not possible?

**Review by mchac15 — APPROVED** [review]
The code seems very good, I don't see any major problems with it but for the duplication of code.

**Review by mchac15 — CHANGES_REQUESTED** [review]
I don't see any major problems but for the duplication of code.
Other than that your code is very well structured and very well written

**Comment by mchac15** [comment]
> I don't see any major problems but for the duplication of code. Other than that your code is very well structured and very well written

I forgot to mention that the Sonar major issue about the MapScreen being too complex will be aproved since this will be done in the refactoring of the MapScreen #137

File: app/src/androidTest/java/ch/hikemate/app/endtoend/EndToEndTest2.kt (hunk 1 above)
Comment (rBergCode) [reply app/src/androidTest/java/ch/hikemate/app/endtoend/EndToEndTest2.kt:50]: Good question! I changed the test to create the user in the UI instead of creating it like this.
I changed it because we need to properly create a profile for this to work.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt (hunk 3 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:399]: I think that is the issue with Sonar.
I will try to refactor it to some kind of function but this might be hard to do.

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt (hunk 2 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:197]: You're right!

**Comment by rBergCode** [comment]
Those tests don't pass because of either End2end or End2End2 test. I don't know why and can't replicate it in local. @Wylwi have you any ideas ?

**Comment by Wylwi** [comment]
Since you're the only one affected by those errors, no sorry, I don't have any idea... It might be your code, but I don't know which part of it

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt (hunk 3 above)
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:399]: Did you try to do it?
Is it too difficult/lengthy to implement?

Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt]: I just did it!

**Comment by rBergCode** [comment]
The tests were fixed by #197. @mchac15 it should be ready now!

File: app/src/main/java/ch/hikemate/app/ui/components/AsyncStateHandler.kt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * content based on the state. If an error occurred, an error message is displayed. If the value is
+ * null, a loading animation is displayed. Otherwise, the content is displayed.
+ *
+ * @param errorMessageIdState The state that holds the error message id
+ * @param actionContentDescriptionStringId The string id for the content description of the action
+ * @param actionOnErrorAction The action to perform when the error action is clicked
+ * @param valueState The state that holds the value to display
+ * @param content The content to display
+ */
+@Composable
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/components/AsyncStateHandler.kt:21]: Comment: This function seems pretty good, I've tested again on my phone and by reviewing the code it seems equivalent to the code you had before, but much better, since there isn't any code repetition.
You also did well by treating the cases whenver the user is not signed in and whenever there is an error this makes debuging much easier.

File: app/src/main/java/ch/hikemate/app/ui/components/HikeCard.kt [hunk 5]
```diff
@@ -71,7 +72,10 @@ fun HikeCard(
     styleProperties: HikeCardStyleProperties = HikeCardStyleProperties(),
     showGraph: Boolean = true,
 ) {
-  val displayMessage = false
+  val displayMessage = !elevationData.isNullOrEmpty() && messageContent != null
+  Log.i(
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/components/HikeCard.kt:76]: Comment: Nice again, of adding a Log (I didn't comment this before.)

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt [hunk 6]
```diff
@@ -185,29 +193,47 @@ fun HikeDetailScreen(
...
-        onZoomIn = { mapView.controller.zoomIn() },
-        onZoomOut = { mapView.controller.zoomOut() },
-        modifier =
-            Modifier.align(Alignment.BottomEnd)
-                .padding(bottom = MapScreen.BOTTOM_SHEET_SCAFFOLD_MID_HEIGHT + 8.dp))
-
-    // Hike Details bottom sheet
-    HikeDetails(detailedRoute, savedHikesViewModel, elevationData)
+  LaunchedEffect(Unit) {
+    if (authViewModel.currentUser.value == null) {
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:197]: Question: Is this case actually possible? If such wouldn't it be more appropriate to show a toast?

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt [hunk 7]
```diff
@@ -185,29 +193,47 @@ fun HikeDetailScreen(
...
+      Log.e("HikeDetailScreen", "User is not signed in")
+      return@LaunchedEffect
+    }
+    profileViewModel.getProfileById(authViewModel.currentUser.value!!.uid)
+  }
+
+  val errorMessageIdState = profileViewModel.errorMessageId.collectAsState()
+  val profileState = profileViewModel.profile.collectAsState()
+
+  AsyncStateHandler(
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:207]: Praise: The function just makes the code much more readable and it reduces the complexity a bit.

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt [hunk 8]
```diff
@@ -210,16 +215,31 @@ fun clearHikesFromMap(mapView: MapView, userLocationMarker: Marker?) {
...
+/**
+ * Properties to initialize the map with. These properties are used to set the initial state of the
+ * map, such as the zoom level, the center, and the limits of the zoom.
+ *
+ * @param mapInitialZoomLevel The initial zoom level of the map
+ * @param mapMaxZoomLevel The maximum zoom level of the map
+ * @param mapMinZoomLevel The minimum zoom level of the map
+ * @param mapInitialCenter The initial center of the map
+ */
+data class MapInitialValues(
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:227]: Comment: I like this improvement, but is there a reason why you changed it? I mean a reason related to the PR?

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt [hunk 9]
```diff
@@ -246,17 +266,23 @@ fun MapScreen(
       // Maximum number of bytes that can be used by the tile file system cache. Default is 600MB.
       tileFileSystemCacheMaxBytes = 600L * 1024L * 1024L
     }
+
+    if (authViewModel.currentUser.value == null) {
```
Comment (mchac15) [inline app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:270]: Same question as in the `HikeDetailsScreen`

**Review by mchac15 — CHANGES_REQUESTED** [review]
The code is very well written and seems to be much improved with the provided function. I added some comments that I didn't do last time.
But overall I don't think there is any major problems and I have tested on my phone looking at the logs and everything seems fine.
Thanks for the changes!

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt (hunk 8 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:227]: It's not really related but I had Sonar complains about to many arguments for MapScreen function

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt (hunk 6 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:197]: It should never happen. This is just a sanity check to prevent NullPointerException

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt (hunk 9 above)
Comment (rBergCode) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:270]: Answered above (will resolve it)

File: app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt (hunk 6 above)
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/HikeDetailScreen.kt:197]: Ok yes that seems more logical

File: app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt (hunk 8 above)
Comment (mchac15) [reply app/src/main/java/ch/hikemate/app/ui/map/MapScreen.kt:227]: Good improvemet

**Review by mchac15 — APPROVED** [review]
Thanks for answering my questions!
You can merge now
